File: app/src/main/java/com/zwstudio/logicgamesandroid/clouds/domain/CloudsGameState.java
package com.zwstudio.logicgamesandroid.clouds.domain;

import com.zwstudio.logicgamesandroid.logicgames.domain.LogicGamesHintState;
import com.zwstudio.logicgamesandroid.logicgames.domain.Position;

import java.util.Arrays;

import fj.F;

/**
 * Created by zwvista on 2016/09/29.
 */

public class CloudsGameState {
    public CloudsGame game;
    public CloudsObject[] objArray;
    public LogicGamesHintState[] row2state;
    public LogicGamesHintState[] col2state;
    public boolean isSolved;

    public Position size() {return game.size;}
    public int rows() {return game.rows();}
    public int cols() {return game.cols();}
    public boolean isValid(int row, int col) { return game.isValid(row, col); }
    public boolean isValid(Position p) {
        return game.isValid(p);
    }

    public CloudsGameState(CloudsGame game) {
        this.game = game;
        objArray = new CloudsObject[rows() * cols()];
        Arrays.fill(objArray, CloudsObject.Empty);
    }

    public CloudsObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public CloudsObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, CloudsObject obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, CloudsObject obj) {
        set(p.row, p.col, obj);
    }

    private void updateIsSolved() {
        isSolved = true;
        for (int r = 0; r < rows(); r++) {
            int n1 = 0, n2 = game.row2hint[r];
            for (int c = 0; c < cols(); c++)
                if (get(r, c) == CloudsObject.Cloud)
                    n1++;
            row2state[r] = n1 < n2 ? LogicGamesHintState.Normal : n1 == n2 ? LogicGamesHintState.Complete : LogicGamesHintState.Error;
            if (n1 != n2) isSolved = false;
        }
        for (int c = 0; c < cols(); c++) {
            int n1 = 0, n2 = game.col2hint[c];
            for (int r = 0; r < rows(); r++)
                if (get(r, c) == CloudsObject.Cloud)
                    n1++;
            col2state[c] = n1 < n2 ? LogicGamesHintState.Normal : n1 == n2 ? LogicGamesHintState.Complete : LogicGamesHintState.Error;
            if (n1 != n2) isSolved = false;
        }
        if (!isSolved) return;
    }

    public boolean setObject(CloudsGameMove move) {
        Position p = move.p;
        CloudsObject objOld = get(p);
        CloudsObject objNew = move.obj;
        set(p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(CloudsMarkerOptions markerOption, CloudsGameMove move) {
        F<CloudsObject, CloudsObject> f = obj -> {
            switch (obj) {
            case Empty:
                return markerOption == CloudsMarkerOptions.MarkerBeforeCloud ?
                        CloudsObject.Marker : CloudsObject.Cloud;
            case Cloud:
                return markerOption == CloudsMarkerOptions.MarkerAfterCloud ?
                        CloudsObject.Marker : CloudsObject.Empty;
            case Marker:
                return markerOption == CloudsMarkerOptions.MarkerBeforeCloud ?
                        CloudsObject.Cloud : CloudsObject.Empty;
            }
            return obj;
        };
        CloudsObject objOld = get(move.p);
        move.obj = f.f(objOld);
        return setObject(move);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/clouds/domain/CloudsGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.clouds.domain

import com.zwstudio.logicpuzzlesandroid.common.domain.*
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState

class CloudsGameState(game: CloudsGame) : CellsGameState<CloudsGame, CloudsGameMove, CloudsGameState>(game) {
    var objArray = Array(rows() * cols()) { CloudsObject.Empty }
    var row2state = Array(rows()) { HintState.Normal }
    var col2state = Array(cols()) { HintState.Normal }

    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: CloudsObject) {objArray[row * cols() + col] = obj}
    operator fun set(p: Position, obj: CloudsObject) {this[p.row, p.col] = obj}

    init {
        updateIsSolved()
    }

    fun setObject(move: CloudsGameMove): Boolean {
        val p = move.p
        if (!isValid(p) || this[p] == move.obj) return false
        this[p] = move.obj
        updateIsSolved()
        return true
    }

    fun switchObject(move: CloudsGameMove): Boolean {
        val markerOption = MarkerOptions.values()[game!!.gdi.markerOption]
        val p = move.p
        if (!isValid(p)) return false
        val o = this[p]
        move.obj = when (o) {
            CloudsObject.Empty -> if (markerOption == MarkerOptions.MarkerFirst) CloudsObject.Marker else CloudsObject.Cloud
            CloudsObject.Cloud -> if (markerOption == MarkerOptions.MarkerLast) CloudsObject.Marker else CloudsObject.Empty
            CloudsObject.Marker -> if (markerOption == MarkerOptions.MarkerFirst) CloudsObject.Cloud else CloudsObject.Empty
            else -> o
        }
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 5/Clouds

        Summary
        Weather Radar Report

        Description
        1. You must find Clouds in the sky.
        2. The hints on the borders tell you how many tiles are covered by Clouds
           in that row or column.
        3. Clouds only appear in rectangular or square areas. Furthermore, their
           width and height is always at least two tiles wide.
        4. Clouds can't touch between themselves, not even diagonally.
    */
    private fun updateIsSolved() {
        val allowedObjectsOnly = game.gdi.isAllowedObjectsOnly
        isSolved = true
        for (r in 0 until rows())
            for (c in 0 until cols())
                if (this[r, c] == CloudsObject.Forbidden)
                    this[r, c] = CloudsObject.Empty
        // 2. The hints on the borders tell you how many tiles are covered by Clouds
        // in that row.
        for (r in 0 until rows()) {
            var n1 = 0
            val n2 = game.row2hint[r]
            for (c in 0 until cols())
                if (this[r, c] == CloudsObject.Cloud)
                    n1++
            row2state[r] = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            if (n1 != n2) isSolved = false
        }
        // 2. The hints on the borders tell you how many tiles are covered by Clouds
        // in that column.
        for (c in 0 until cols()) {
            var n1 = 0
            val n2 = game!!.col2hint[c]
            for (r in 0 until rows())
                if (this[r, c] == CloudsObject.Cloud)
                    n1++
            col2state[c] = if (n1 < n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            if (n1 != n2) isSolved = false
        }
        for (r in 0 until rows())
            for (c in 0 until cols()) {
                val o = this[r, c]
                if ((o == CloudsObject.Empty || o == CloudsObject.Marker) && allowedObjectsOnly && (row2state[r] != HintState.Normal || col2state[c] != HintState.Normal))
                    this[r, c] = CloudsObject.Forbidden
            }
        if (!isSolved) return
        val g = Graph()
        val pos2node = mutableMapOf<Position, Node>()
        for (r in 0 until rows())
            for (c in 0 until cols()) {
                val p = Position(r, c)
                if (this[p] != CloudsObject.Cloud) continue
                val node = Node(p.toString())
                g.addNode(node)
                pos2node[p] = node
            }
        for (p in pos2node.keys)
            for (os in CloudsGame.offset) {
                val p2 = p.add(os)
                if (pos2node.containsKey(p2))
                    g.connectNode(pos2node[p], pos2node[p2])
            }
        while (pos2node.isNotEmpty()) {
            g.rootNode = pos2node.values.first()
            val nodeList = g.bfs()
            var r2 = 0
            var r1 = rows()
            var c2 = 0
            var c1 = cols()
            for (node in nodeList) {
                val p = pos2node.filterValues { it == node }.keys.first()
                pos2node.remove(p)
                if (r2 < p.row) r2 = p.row
                if (r1 > p.row) r1 = p.row
                if (c2 < p.col) c2 = p.col
                if (c1 > p.col) c1 = p.col
            }
            val rs = r2 - r1 + 1
            val cs = c2 - c1 + 1
            // 3. Clouds only appear in rectangular or square areas. Furthermore, their
            // width and height is always at least two tiles wide.
            if (!(rs >= 2 && cs >= 2 && rs * cs == nodeList.size)) {
                isSolved = false
                return
            }
        }
    }
}